src/VAC/VectorAnimationComplex/InfillPattern.cpp

```diff
@@ -322,6 +327,120 @@ QVector<QPointF> connectInfillAlongInset(const QVector<QVector<QPair<QPointF, in
+    }
+
+    // Start from closest segment and travel across polygon, piecing together new polygon without intersections
+    auto segmentIndex = std::get<1>(closestSegmentToInteriorPoint.value());
+    const auto endSegmentIndex = segmentIndex;
+    const auto polygonSize = polygon.size();
+
+    auto startPoint = std::get<0>(closestSegmentToInteriorPoint.value());// polygon[segmentIndex % polygonSize];
+    do {
+        // Check all lines against this line to find intersections
+        const auto endPoint = polygon[(segmentIndex + 1) % polygonSize];
+        const auto compareDirectionX = roundDouble(endPoint.x()) - roundDouble(startPoint.x()) > 0;
+        const auto compareDirectionY = roundDouble(endPoint.y()) - roundDouble(startPoint.y()) > 0;
+        auto innerSegmentIndex = (segmentIndex + 1) % polygonSize;
+        // Find self intersections
+        QVector<std::tuple<QPointF, int>> selfIntersections;
+        do {
+            if (const auto intersection = segmentSegmentIntersection(startPoint,
+                                                                     endPoint,
+                                                                     polygon[innerSegmentIndex % polygonSize],
+                                                                     polygon[(innerSegmentIndex + 1) % polygonSize])) {
+                // Make sure the point is in right direction
+                const auto directionX = roundDouble(intersection->x()) - roundDouble(startPoint.x()) > 0;
+                const auto directionY = roundDouble(intersection->y()) - roundDouble(startPoint.y()) > 0;
+                if (intersection.value() != startPoint && directionX == compareDirectionX && directionY == compareDirectionY) {
+                    selfIntersections.append({intersection.value(), innerSegmentIndex});
+                }
+            }
+            innerSegmentIndex = (innerSegmentIndex + 1) % polygonSize;
+        } while(innerSegmentIndex != segmentIndex);
+
+        // Find closest intersection segment
+        std::optional<std::tuple<QPointF, int>> closestSegmentToSegment{};
+        auto smallestDistance = maxNumber;
+        for (const auto &point : selfIntersections){
+            if (const auto distanceSquared = pointToPointDistanceSquared(startPoint, std::get<0>(point)); distanceSquared < smallestDistance){
+                smallestDistance = distanceSquared;
+                closestSegmentToSegment = point;
+            }
+        }
+
+        // Append correct point and increase segmentIndex
+        if (!closestSegmentToSegment) {
+            qWarning() << "Failed to find closestSegmentToSegment";
+            break;
+        }
+        const auto intersectionPoint = std::get<0>(closestSegmentToSegment.value());
+        polygonWithoutIntersections << intersectionPoint;
+        segmentIndex = std::get<1>(closestSegmentToSegment.value());
+        startPoint = intersectionPoint;
+    } while (segmentIndex != endSegmentIndex);
```

Comment:
(segmentIndex < endSegmentIndex) could be more safe

Reply:
I do think this is fine since it will always loop back to that value. (segmentIndex < endSegmentIndex) doesn't work because segmentIndex is often higher than endSegmentIndex but should continue except for when they become equal and it will break either way.
(The best solution would have been to use some sort of linked list for what I'm doing. But I couldn't get that to work properly when I tried it. Maybe that's a future refactoring...)